### ISSUE TITLE
Deps not upgraded when setting up Poetry's venv

### ISSUE THREAD

**Issue by kadams54** [python-poetry/install.python-poetry.org#100]
This was an engaging situation we ran into yesterday, so bear with me because 1) I'm going to try to provide lots of details, and 2) it's a bit convoluted.

**First the TLDR:** not setting the `upgrade_deps` flag on `venv.EnvBuilder` means Poetry's virtualenv (built on install) may have dependencies with security vulnerabilities.

**The long version:** we use [Trivy](https://trivy.dev/) to scan our Docker containers as part of CI. Yesterday Trivy started failing our builds due to [CVE-2022-40897](CVE-2022-40897). We upgraded the `setuptools` in the container's system Python, but it still failed, as it still found an old version of `setuptools` in this path:

```
/root/.local/share/pypoetry/venv/lib/python3.9/site-packages/setuptools-44.0.0.dist-info/METADATA
```

We eventually figured out that this is the virtualenv covered by the comments at the top of the install script. What we discovered is that the `venv` module has its own embedded versions of `pip` and `setuptools`, separate from the system Python. Consequently, even if we upgraded the system Python, we'd need to force `venv` to _also_ upgrade these embedded deps.

We hacked around the issue by, post-Poetry-install, activating the virtualenv, installing the up-to-date `setuptools`, and then deactivating. From our `Dockerfile`:

```
RUN source /root/.local/share/pypoetry/venv/bin/activate && \
    pip3 install --upgrade setuptools && \
    deactivate
```

It works, but it's a hack, so I went digging for a better/more long-term fix. My first attempt was to upgrade the embedded wheels pre-Poetry-install, using the `virtualenv` module and its `--upgrade-embed-wheels` switch. Unfortunately, that doesn't seem to work, as `venv` and `virtualenv` are two different beasts. Consequently, the fix needs to happen on the `venv` invocation, which takes place in the Poetry install script, when it [invokes venv.EnvBuilder](https://github.com/python-poetry/install.python-poetry.org/blob/main/install-poetry.py#L309). Hence this bug.

I'll have a corresponding PR with the fix posted shortly.

**Comment by neersighted** [comment]
Your approach is not going to work -- I suppose we could add a `pip install -U pip setuptools wheel` to the installer after the virtual environment is created, though I'm not sure we want to do so unconditionally; the performance regression will likely be painful for many users.

The CVE you have provided is inert in this scenario as we only search PyPI (and that copy of setuptools is only used by the copy of pip in Poetry's environment); if you can't trust PyPI we have much bigger problems.

Unconditionally upgrading feels like cutting off our nose to spite our face, especially when you consider that any non-inert RCE in pip/setuptools will have to be invoked as part of the upgrade.

**Comment by kadams54** [comment]
Yeah, I ran into that issue with `upgrade_deps` in the associated PR. I ended up going the route you mentioned. I think there's still value in upgrading from a "future-proofing against security patches" perspective. I also think there's precedent in the upgrade the script is _already_ doing for `pip`.

What I don't know—and you raise a good point here—is what the additional overhead is for upgrading `setuptools`. I'll see if I can get at least some anecdotal data.
